pallets/auctions/src/lib.rs

```diff
@@ -0,0 +1,832 @@
+			common_data.token.0,
+			common_data.token.1,
+		)?;
+
+		Ok(())
+	}
+
+	///
+	/// Validates certain aspects relevant to the bid action
+	///
+	fn validate_bid(
+		bidder: &<T>::AccountId,
+		common_auction_data: &CommonAuctionData<T>,
+		bid: &Bid<T>,
+	) -> DispatchResult {
+		let block_number = <frame_system::Pallet<T>>::block_number();
+		ensure!(bidder != &common_auction_data.owner, Error::<T>::CannotBidOnOwnAuction);
+		ensure!(block_number > common_auction_data.start, Error::<T>::AuctionNotStarted);
+		ensure!(
+			block_number < common_auction_data.end,
+			Error::<T>::AuctionEndTimeReached
+		);
+		ensure!(
+			bid.amount >= common_auction_data.next_bid_min,
+			Error::<T>::InvalidBidPrice
+		);
+
+		if let Some(current_bid) = &common_auction_data.last_bid {
+			ensure!(bid.amount > current_bid.1, Error::<T>::InvalidBidPrice);
+		} else {
+			ensure!(!bid.amount.is_zero(), Error::<T>::InvalidBidPrice);
+		}
+
+		Ok(())
+	}
+
+	///
+	/// Validates certain aspects relevant to the close action
+	///
+	fn validate_close(common_auction_data: &CommonAuctionData<T>) -> DispatchResult {
+		ensure!(!common_auction_data.closed, Error::<T>::AuctionClosed);
+		ensure!(
+			Pallet::is_auction_ended(common_auction_data),
+			Error::<T>::AuctionEndTimeNotReached
+		);
+
+		Ok(())
+	}
+
+	///
+	/// Validates certain aspects relevant to the claim action
+	///
+	fn validate_claim(common_auction_data: &CommonAuctionData<T>) -> DispatchResult {
+		ensure!(
+			Pallet::<T>::is_auction_ended(common_auction_data),
+			Error::<T>::AuctionEndTimeNotReached
+		);
+		ensure!(
+			common_auction_data.closed,
+			Error::<T>::CloseAuctionBeforeClaimingReservedAmounts
+		);
+
+		Ok(())
+	}
+
+	///
+	/// Helper function for handling a claim
+	/// 
+	/// It transfers the reserved amount to the bidder after which destroys the record from store
+	///
+	fn handle_claim(bidder: T::AccountId, auction_id: T::AuctionId, amount: BalanceOf<T>) -> DispatchResult {
+		<<T as crate::Config>::Currency as Currency<T::AccountId>>::transfer(
+			&Pallet::<T>::get_auction_subaccount_id(auction_id),
+			&bidder,
+			amount,
+			ExistenceRequirement::AllowDeath,
+		)?;
+
+		<ReservedAmounts<T>>::remove(bidder, auction_id);
+
+		Ok(())
+	}
+
+	fn set_next_bid_min(common_auction_data: &mut CommonAuctionData<T>, amount: BalanceOf<T>) -> DispatchResult {
+		let bid_step = Permill::from_percent(<T as crate::Config>::BidStepPerc::get()).mul_floor(amount);
+		common_auction_data.next_bid_min = amount.checked_add(&bid_step).ok_or(Error::<T>::BidOverflow)?;
+
+		Ok(())
+	}
+
+	///
+	/// Helper function which extends auction end time if necessary to prevent auction sniping
+	/// 
+	fn avoid_auction_sniping(common_auction_data: &mut CommonAuctionData<T>) -> DispatchResult {
+		let block_number = <frame_system::Pallet<T>>::block_number();
+		let time_left = common_auction_data
+			.end
+			.checked_sub(&block_number)
+			.ok_or(Error::<T>::TimeUnderflow)?;
```

Comment:
I would predict a test for this `TimeUnderflow` will fail a `assert_noop!` (see remarks about atomicity).